File: src/simulation/space/resource/dependency/ConsumeDependency.java
package simulation.space.resource.dependency;

import simulation.space.resource.tag.labeler.ResourceLabeler;
import simulation.space.tile.Tile;
import simulation.space.resource.Resource;

import java.util.HashSet;
import java.util.Set;

public class ConsumeDependency extends LabelerDependency {
    public Set<String> lastConsumed = new HashSet<>();
    int currentAmount = 0;

    public ConsumeDependency(
            double deprivationCoefficient,
            boolean isNecessary,
            double amount,
            ResourceLabeler goodResource) {
        super(deprivationCoefficient, isNecessary, amount, goodResource);
    }

    @Override
    double satisfaction(Tile tile, Resource resource) {
        if (currentAmount < 0) currentAmount = 0;
        double result;
        double neededAmount = amount * (resource == null ? 1 : resource.getAmount());
        if (currentAmount < neededAmount) {
            for (Resource res : tile.getAccessibleResources()) {
                if (res.equals(resource)) {
                    continue;
                }
                if (isResourceDependency(res)) {
                    Resource part = res.getPart(partByResource(res, neededAmount - currentAmount));
                    if (part.isNotEmpty()) {
                        lastConsumed.add(part.getFullName());
                        currentAmount += part.getAmount() * oneResourceWorth(res);
                    }
                    part.destroy();
                    if (currentAmount >= neededAmount) {
                        break;
                    }
                }
            }
        }
        result = Math.min(((double) currentAmount) / neededAmount, 1);
        if (currentAmount >= neededAmount) {
            currentAmount -= neededAmount;
        }
        return result;
    }

    @Override
    public boolean isPositive() {
        return true;
    }
}

File: src/simulation/space/resource/dependency/ConsumeDependency.kt
package simulation.space.resource.dependency

import simulation.space.resource.Resource
import simulation.space.resource.tag.labeler.ResourceLabeler
import simulation.space.tile.Tile
import java.util.*
import kotlin.math.min


class ConsumeDependency(
        deprivationCoefficient: Double,
        isNecessary: Boolean,
        amount: Double,
        goodResource: ResourceLabeler?) : LabelerDependency(deprivationCoefficient, isNecessary, amount, goodResource
) {
    var lastConsumed: MutableSet<String> = HashSet()

    var currentAmount = 0

    override fun satisfaction(tile: Tile, resource: Resource): Double {
        if (currentAmount < 0)
            currentAmount = 0

        val result: Double
        val neededAmount = amount * resource.amount

        if (currentAmount < neededAmount) {
            for (list in tile.accessibleResources)
                for (res in list) {
                    if (res == resource)
                        continue

                    if (isResourceDependency(res)) {
                        val part = res.getPart(partByResource(res, neededAmount - currentAmount))
                        if (part.isNotEmpty) {
                            lastConsumed.add(part.fullName)
                            currentAmount += part.amount * oneResourceWorth(res)
                        }

                        part.destroy()
                        if (currentAmount >= neededAmount)
                            break
                    }
                }
        }

        result = min(currentAmount.toDouble() / neededAmount, 1.0)

        if (currentAmount >= neededAmount)
            currentAmount -= neededAmount.toInt()

        return result
    }

    override val isPositive: Boolean
        get() = true
}
